# Entregable nº2 , ML clasificación 'churn' Telecom - PREDICT
## Ricardo Stocco, Maria Vidondo, Victor Borrajo

In [1]:
pip install dill

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 110.5/110.5 KB 5.8 MB/s eta 0:00:00


In [2]:
import pandas as pd     # Manejo de dataframes
import numpy as np      # Calculos matriciales
import matplotlib.pyplot as plt    # Visualizacion
import seaborn as sns              # Visualizacion
from google.colab import drive     # Drive en Colab

pd.set_option('display.max_columns',100)   # para visualizar mas columnas en un df

import warnings
warnings.filterwarnings("ignore")  # inhabilita warnings
import dill as pickle

In [3]:
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [4]:
PATH= '/content/drive/MyDrive/DSC1122 –Entregable 2-Borrajo_Stocco_Vidondo/'

### Importarmos el dataset para realizar la predicción y el modelo escogido.

In [34]:
X_pred= pd.read_csv(PATH + 'telecom_churn_PREDICT.csv')
X_pred_customer_ID= pd.read_csv(PATH + 'telecom_churn_PREDICT.csv') #Cargamos dos veces el dataset para obtener el Customer ID
Customer_ID_index = X_pred_customer_ID[['Customer_ID']]

In [6]:
best_model= pickle.load(open(PATH + 'model.pkl','rb'))

In [7]:
X_pred.head()

rev      mou    totmrc      da  ovrmou  ovrrev  vceovr  datovr  \
0   30.8350   136.75   29.9900  0.2475    1.25   0.500     0.5   0.000   
1   35.8475   352.75   24.2700  0.4950   23.25   9.285     8.7   0.585   
2   30.3275   241.50   39.9900  0.0000    0.00   0.000     0.0   0.000   
3  154.6925  2297.00  149.9900  4.7025    0.00   0.000     0.0   0.000   
4  156.0050   542.25   48.9475  0.7425   70.75  28.295    28.1   0.195   

     roam  change_mou  change_rev   drop_vce  drop_dat   blck_vce  blck_dat  \
0  0.0975       48.25      1.1550   2.000000  0.000000  10.333333       0.0   
1  1.8000     -352.75    -35.8475   9.000000  0.000000   0.000000       0.0   
2  0.0000      -86.50      0.8625   2.000000  0.000000   4.666667       0.0   
3  0.0000      119.00      0.2475  20.333333  0.000000   5.000000       0.0   
4  0.0000      195.75     40.0150   5.666667  0.333333   0.666667       0.0   

     unan_vce  unan_dat     plcd_vce  plcd_dat    recv_vce  recv_sms  \
0   13.333333       0.0    66.000000  0.000000    3.333333       0.0   
1   10.000000       0.0    82.666667  1.000000   31.000000       0.0   
2   20.666667       0.0    85.333333  0.000000   26.666667       0.0   
3  329.333333       0.0  1067.000000  0.000000  628.000000       0.0   
4   46.333333       0.0   167.666667  0.666667   11.333333       0.0   

     comp_vce  comp_dat   custcare   ccrndmou     cc_mou    inonemin  \
0   40.666667  0.000000   1.333333   1.333333   0.896667    0.666667   
1   65.333333  1.000000   3.666667   9.000000   5.890000   18.333333   
2   59.333333  0.000000   4.333333   6.333333   4.466667   17.000000   
3  723.000000  0.000000   0.333333   0.666667   0.376667  408.000000   
4  115.000000  0.333333  21.666667  48.000000  31.346667   10.000000   

   threeway    mou_cvce  mou_cdat    mou_rvce  owylis_vce  mouowylisv  \
0  0.333333   71.746667  0.000000    5.406667    1.333333    0.713333   
1  1.333333   98.543333  1.023333   83.666667   15.666667   12.683333   
2  1.666667  143.530000  0.000000   27.373333    9.666667    3.760000   
3  0.000000  769.136667  0.000000  679.020000  271.000000  167.090000   
4  0.333333  343.066667  0.010000   11.243333   13.000000   14.563333   

   iwylis_vce  mouiwylisv    peak_vce  peak_dat    mou_peav  mou_pead  \
0    0.333333    0.473333   30.000000       0.0   54.500000       0.0   
1    1.333333   14.286667   38.000000       0.0  106.926667       0.0   
2   19.666667   21.736667   48.666667       0.0   76.356667       0.0   
3   68.666667  150.656667  723.000000       0.0  961.736667       0.0   
4    1.333333    2.100000   91.000000       0.0  205.216667       0.0   

      opk_vce   opk_dat    mou_opkv  mou_opkd   drop_blk      attempt  \
0   13.666667  0.000000   22.653333  0.000000  12.333333    66.000000   
1   43.666667  1.000000   75.283333  1.023333   9.000000    83.666667   
2   36.000000  0.000000   94.543333  0.000000   6.666667    85.333333   
3  346.000000  0.000000  486.420000  0.000000  25.333333  1067.000000   
4   31.000000  0.333333  149.096667  0.010000   6.666667   168.333333   

     complete  callfwdv  callwait  churn  months  uniqsubs  actvsubs new_cell  \
0   40.666667       0.0  0.000000      1       7         1         1        U   
1   66.333333       0.0  1.333333      0       7         1         1        U   
2   59.333333       0.0  0.333333      0       9         2         2        Y   
3  723.000000       0.0  8.333333      1      11         1         1        U   
4  115.333333       0.0  0.666667      0       9         1         1        U   

  crclscod asl_flag  totcalls   totmou   totrev   adjrev   adjmou  adjqty  \
0       AA        N       393    958.0   248.46   182.94    958.0     393   
1       EA        Y      1641   4011.0   391.87   305.59   3998.0    1634   
2        A        N       780   1889.0   312.91   219.45   1772.0     734   
3       AA        N      9018  18478.0  2217.05  2131.17  18478.0    9018   
4       CA        Y       791   2

In [8]:
#Comprobamos que el dataset para la predicción tiene nulos, por lo que usaremos los mismos pasos del pre-processing
X_pred.isnull().sum()

rev            223
mou            223
totmrc         223
da             223
ovrmou         223
              ... 
kid11_15       281
kid16_17       281
creditcd       281
eqpdays          0
Customer_ID      0
Length: 100, dtype: int64

In [9]:
X_pred.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 100 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   rev               9777 non-null   float64
 1   mou               9777 non-null   float64
 2   totmrc            9777 non-null   float64
 3   da                9777 non-null   float64
 4   ovrmou            9777 non-null   float64
 5   ovrrev            9777 non-null   float64
 6   vceovr            9777 non-null   float64
 7   datovr            9777 non-null   float64
 8   roam              9777 non-null   float64
 9   change_mou        9704 non-null   float64
 10  change_rev        9704 non-null   float64
 11  drop_vce          10000 non-null  float64
 12  drop_dat          10000 non-null  float64
 13  blck_vce          10000 non-null  float64
 14  blck_dat          10000 non-null  float64
 15  unan_vce          10000 non-null  float64
 16  unan_dat          10000 non-null  float6

# PRE-PROCESSING

In [10]:
#Eliminamos el customer id y churn (ya que esta columna es el valor que predeciremos)
del(X_pred['Customer_ID'])
del(X_pred['churn'])

In [11]:
#Comprobamos que no hay duplicados
X_pred.duplicated().sum()

0

Importamos los diccionarios, las funciones utilizadas en el preprocessing y las lista de columnas eliminadas por baja varianza y alta correlación:

In [12]:
dict_input_media= pickle.load(open(PATH +'dict_input_media.pkl','rb'))
dict_input_moda= pickle.load(open(PATH +'dict_input_moda.pkl','rb'))
cols_to_remove= pickle.load(open(PATH +'cols_to_remove.pkl','rb'))


In [13]:
imputacion_nulos= pickle.load(open(PATH +'imputacion_nulos.pkl','rb'))
del_valores_negativos= pickle.load(open(PATH +'del_valores_negativos.pkl','rb'))
new_column_kids= pickle.load(open(PATH +'new_column_kids.pkl','rb'))
cat_columns_to_num= pickle.load(open(PATH +'cat_columns_to_num.pkl','rb'))
num_to_bool= pickle.load(open(PATH +'num_to_bool.pkl','rb'))

In [14]:
#Aplicamos el mismo preprocesamiento hecho al dataset train-test
def prepro(df,dict_input_media,dict_input_moda):
    imputacion_nulos(df, dict_input_media, dict_input_moda)
    del_valores_negativos(df)
    new_column_kids(df)
    cat_columns_to_num(df)
    num_to_bool(df)

In [15]:
prepro(X_pred,dict_input_media, dict_input_moda)

In [16]:
#Comprobamos que no hay nulos en el dataset para la predicción.
X_pred.isnull().sum().sum()

0

In [17]:
X_pred.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 94 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   rev               10000 non-null  float64
 1   mou               10000 non-null  float64
 2   totmrc            10000 non-null  float64
 3   da                10000 non-null  float64
 4   ovrmou            10000 non-null  float64
 5   ovrrev            10000 non-null  float64
 6   vceovr            10000 non-null  float64
 7   datovr            10000 non-null  float64
 8   roam              10000 non-null  float64
 9   change_mou        10000 non-null  float64
 10  change_rev        10000 non-null  float64
 11  drop_vce          10000 non-null  float64
 12  drop_dat          10000 non-null  float64
 13  blck_vce          10000 non-null  float64
 14  blck_dat          10000 non-null  float64
 15  unan_vce          10000 non-null  float64
 16  unan_dat          10000 non-null  float64

Tenemos un dataframe para predecir sin nulos, aplicamos OneHotEncoding

In [18]:
X_pred_enc = pd.get_dummies(X_pred, columns=['new_cell','crclscod','asl_flag','prizm_social_one','area','dualband','refurb_new','hnd_webcap','ownrent','dwlltype','marital','infobase','ethnic',
 'creditcd'])

A este dataframe le tenemos que eliminar las col. correlacionadas y de baja varianza

In [19]:
for i in X_pred_enc.columns:
    if i in cols_to_remove:
        del(X_pred_enc[i])

Tenemos que asegurarnos que las columnas con las que se ha entrenado el modelo son las mismas que tenemos en el dataset para la predicción:

In [20]:
features_model =  list(best_model.feature_names_in_)
print('Estas son las variables que tenemos que conseguir en el preprocessing para que el modelo funcione:\n', features_model)

Estas son las variables que tenemos que conseguir en el preprocessing para que el modelo funcione:
 ['rev', 'mou', 'totmrc', 'da', 'vceovr', 'datovr', 'roam', 'change_mou', 'change_rev', 'drop_vce', 'drop_dat', 'blck_vce', 'blck_dat', 'unan_vce', 'unan_dat', 'plcd_dat', 'recv_vce', 'recv_sms', 'custcare', 'ccrndmou', 'threeway', 'mou_cvce', 'mou_cdat', 'mou_rvce', 'owylis_vce', 'mouowylisv', 'iwylis_vce', 'mouiwylisv', 'peak_vce', 'peak_dat', 'mou_peav', 'mou_pead', 'opk_vce', 'opk_dat', 'mou_opkv', 'drop_blk', 'complete', 'callfwdv', 'callwait', 'months', 'uniqsubs', 'actvsubs', 'adjrev', 'adjmou', 'adjqty', 'avgrev', 'avgmou', 'avgqty', 'avg3qty', 'avg3rev', 'avg6rev', 'hnd_price', 'phones', 'models', 'truck', 'rv', 'lor', 'adults', 'income', 'numbcars', 'HHstatin', 'dwllsize', 'forgntvl', 'eqpdays', 'n_kids', 'new_cell_N', 'new_cell_U', 'new_cell_Y', 'crclscod_A', 'crclscod_AA', 'crclscod_B', 'crclscod_BA', 'crclscod_C', 'crclscod_CA', 'crclscod_DA', 'crclscod_EA', 'crclscod_ZA', 'a

In [21]:
features_test = list(X_pred_enc.columns)

print('Columnas en dataset:',len(features_test))
print('Variables en modelos:',len(features_model))
print('¿Match?:', features_model == features_test)



Columnas en dataset: 129
Variables en modelos: 127
¿Match?: False


Comprobamos que columnas tenemos en el dataset pero que no son necesarias para la predicción.

In [22]:
no_necesitamos=[]

for i in features_test:
    if i in features_model:
        continue
    else:
        no_necesitamos.append(i)
no_necesitamos

['infobase_N', 'ethnic_C']

Tenemos dos columnas "nuevas", dos columnas que surgen de OneHotEncoding porque tenemos categorías nuevas que no habiamos visto, por lo tanto las borraremos.

In [23]:
del(X_pred_enc['infobase_N'])
del(X_pred_enc['ethnic_C'])

In [24]:
features_test = list(X_pred_enc.columns)

print('Columnas en dataset:',len(features_test))
print('Variables en modelos:',len(features_model))
print('¿Match?:', features_model == features_test)

Columnas en dataset: 127
Variables en modelos: 127
¿Match?: True


Como el modelo utilizado es un modelo de reglas no es necesario escalar los datos. 

In [25]:
best_model

HistGradientBoostingClassifier(l2_regularization=2.5, learning_rate=0.7,
                               loss='auto', max_bins=25, max_depth=2,
                               max_iter=150, max_leaf_nodes=10,
                               min_samples_leaf=25)

## Predicción:

In [26]:
predictions = best_model.predict(X_pred_enc)
predictions

array([0, 0, 0, ..., 0, 0, 0])

In [27]:
#Guardamos nuestras predicción en un dataframe.
predictions = pd.DataFrame(predictions, columns=['churn'], index = X_pred_enc.index)

In [28]:
predictions

churn
0         0
1         0
2         0
3         0
4         0
...     ...
9995      1
9996      1
9997      0
9998      0
9999      0

[10000 rows x 1 columns]

In [37]:
df_final_prediction= pd.merge(
    left= Customer_ID_index,
    right= predictions,
    left_index=True,
    right_index=True
)

In [38]:
df_final_prediction #Hacemos merge del Customer_ID y churn para identificar los clientes que van a darse de baja del servicio.

Customer_ID  churn
0         1090001      0
1         1090002      0
2         1090003      0
3         1090004      0
4         1090005      0
...           ...    ...
9995      1099996      1
9996      1099997      1
9997      1099998      0
9998      1099999      0
9999      1100000      0

[10000 rows x 2 columns]

In [30]:
predictions['churn'].value_counts()


0    7315
1    2685
Name: churn, dtype: int64

En la predicción se observan 2685 clientes que hacen churn vs 7315 que se quedan en la compañia.

In [29]:
predictions.to_csv(PATH + 'prediction.csv')

# CONCLUSIONES FINALES: 

Con los resultados obtenidos en la interpretabilidad del modelo podemos recomendar: 
- Focalizar campañas de marketing de renovación de equipamiento a clientes con equipamiento con una antiguedad cercana a cumplir un año.
- Ofrecer nuevos planes de consumo a los clientes que superan el uso de consumo de voz. 
- Seguimiento más detallado a los nuevos clientes ya que en este periodo tienen más posibilidades de darse de baja. 

### COMENTARIOS ADICIONALES: 
- Hemos probado diferentes estrategias de imputación de nulos así como creación de nuevas variables pero en ninguno de los casos había una diferencia considerable en el Score final, pero el tiempo de ejecución del notebook aumentaba, por lo que hemos dedicido tomar una variante más simple pero más rápida y ágil de ejecutar.